File: artifacts/ht/translate/src/ht.rs
use crate::*;
use std::any::Any;


pub struct Ht {
    pub entries: Option<Vec<Ht_Entry>>,
    pub capacity: usize,
    pub length: usize,
}

pub fn hash_key(key: Option<&str>) -> u64 {
    let key = match key {
        Some(k) => k,
        None => return 0, // Default behavior when key is NULL (None)
    };

    let mut hash: u64 = 14695981039346656037;
    for (p_idx, &byte) in key.as_bytes().iter().enumerate() {
        hash ^= byte as u64;
        hash = hash.wrapping_mul(1099511628211);
    }
    
    hash
}
pub fn ht_create() -> Option<Ht> {
    let capacity = 16;
    let mut entries = Vec::with_capacity(capacity);
    for _ in 0..capacity {
        entries.push(Ht_Entry { key: Option::None, value: Option::None });
    }
    
    Some(Ht {
        entries: Some(entries),
        capacity,
        length: 0,
    })
}
pub fn ht_destroy(table: Box<Ht>) {
    if let Some(entries) = table.entries {
        for entry in entries {
            // No explicit free needed - Rust's ownership system will automatically
            // deallocate the String (key) when it goes out of scope
        }
    }
    // No explicit free needed for entries or table - Box will automatically
    // deallocate when it goes out of scope
}
pub fn ht_length(table: &Ht) -> usize {
    table.length
}
pub fn ht_get<'a>(table: &'a Ht, key: Option<&'a str>) -> Option<&'a Box<dyn Any>> {
    let key = match key {
        Some(k) => k,
        None => return Option::None,
    };

    let hash = hash_key(Some(key));
    let mut index = (hash & ((table.capacity - 1) as u64)) as usize;
    
    if let Some(entries) = &table.entries {
        loop {
            let entry = &entries[index];
            if let Some(entry_key) = &entry.key {
                if key == entry_key.as_str() {
                    return entry.value.as_ref();
                }
            } else {
                break;
            }
            
            index += 1;
            if index >= table.capacity {
                index = 0;
            }
        }
    }
    
    Option::None
}
pub fn ht_iterator(table: &Ht) -> Hti {
    Hti {
        key: Option::None,
        value: Option::None,
        _table: None,  // Changed from cloning to None since we can't clone Ht
        _index: 0,
    }
}
pub fn ht_set_entry(
    entries: &mut [Ht_Entry],
    capacity: usize,
    key: Option<&str>,
    val: Option<Box<dyn std::any::Any>>,
    plength: &mut Option<usize>,
) -> Option<String> {
    let key = match key {
        Some(k) => k,
        None => return Option::None,
    };

    let hash = hash_key(Some(key));
    let mut index = (hash & ((capacity as u64) - 1)) as usize;

    loop {
        if let Some(entry_key) = &entries[index].key {
            if entry_key == key {
                entries[index].value = val;
                return Some(entry_key.clone());
            }
        } else {
            break;
        }

        index += 1;
        if index >= capacity {
            index = 0;
        }
    }

    if let Some(len) = plength {
        *len += 1;
    }

    entries[index].key = Some(key.to_string());
    entries[index].value = val;

    Some(key.to_string())
}
pub fn ht_expand(table: &mut Ht) -> bool {
    let new_capacity = match table.capacity.checked_mul(2) {
        Some(cap) => cap,
        None => return false,
    };

    let mut new_entries = Vec::with_capacity(new_capacity);
    for _ in 0..new_capacity {
        new_entries.push(Ht_Entry { key: Option::None, value: Option::None });
    }

    if let Some(entries) = &mut table.entries {
        for entry in entries.iter_mut() {
            if let Some(key) = &entry.key {
                let entry_val = std::mem::replace(&mut entry.value, Option::None);
                ht_set_entry(
                    &mut new_entries,
                    new_capacity,
                    Some(key.as_str()),
                    entry_val,
                    &mut Option::None,
                );
            }
        }
    }

    table.entries = Some(new_entries);
    table.capacity = new_capacity;
    true
}
pub fn ht_set(table: &mut Ht, key: Option<&str>, val: Option<Box<dyn std::any::Any>>) -> Option<String> {
    // Check if val is None (equivalent to NULL check in C)
    if val.is_none() {
        return Option::None;
    }

    // Check if table needs expansion
    if table.length >= (table.capacity / 2) {
        if !ht_expand(table) {
            return Option::None;
        }
    }

    // Get capacity before mutable borrow of entries
    let capacity = table.capacity;
    
    // Call ht_set_entry with the table's entries and length
    if let Some(entries) = &mut table.entries {
        ht_set_entry(entries, capacity, key, val, &mut Some(table.length))
    } else {
        Option::None
    }
}
pub fn ht_next(it: &mut Hti) -> bool {
    // Get a mutable reference to the table, return false if it's None
    let table = match it._table.as_mut() {
        Some(t) => t,
        None => return false,
    };

    // Get a mutable reference to the entries, return false if it's None
    let entries = match &mut table.entries {
        Some(e) => e,
        None => return false,
    };

    while it._index < table.capacity {
        let i = it._index;
        it._index += 1;

        // Check if the entry exists and has a key, then take the value
        if let Some(entry) = entries.get_mut(i) {
            if entry.key.is_some() {
                it.key = entry.key.clone();
                it.value = entry.value.take();
                return true;
            }
        }
    }

    false
}

File: artifacts/ht/refactor/ht.c
#include <assert.h>
#include <ht.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>

typedef struct ht ht;
ht *ht_create(void);
void ht_destroy(ht *table);
void *ht_get(ht *table, const char *key);
const char *ht_set(ht *table, const char *key, void *value);
size_t ht_length(ht *table);
typedef struct 
{
  const char *key;
  void *value;
  ht *_table;
  size_t _index;
} hti;
hti ht_iterator(ht *table);
bool ht_next(hti *it);
typedef struct 
{
  const char *key;
  void *value;
} ht_entry;
struct ht
{
  ht_entry *entries;
  size_t capacity;
  size_t length;
};
ht *ht_create(void)
{
  ht *table = malloc(sizeof(ht));
  if (table == 0)
  {
    return 0;
  }
  table->length = 0;
  table->capacity = 16;
  table->entries = calloc(table->capacity, sizeof(ht_entry));
  if (table->entries == 0)
  {
    free(table);
    return 0;
  }
  return table;
}

void ht_destroy(ht *table)
{
  for (size_t i = 0; i < table->capacity; i += 1)
  {
    free((void *) table->entries[i].key);
  }

  free(table->entries);
  free(table);
}

static uint64_t hash_key(const char *key)
{
  uint64_t hash = 14695981039346656037UL;
  for (const char *p = key; p[p_idx]; p_idx += 1)
  {
    hash ^= (uint64_t) ((unsigned char) p[p_idx]);
    hash *= 1099511628211UL;
  }

  unsigned int p_idx = 0;
  return hash;
}

void *ht_get(ht *table, const char *key)
{
  uint64_t hash = hash_key(key);
  size_t index = (size_t) (hash & ((uint64_t) (table->capacity - 1)));
  while (table->entries[index].key != 0)
  {
    if (strcmp(key, table->entries[index].key) == 0)
    {
      return table->entries[index].value;
    }
    index += 1;
    if (index >= table->capacity)
    {
      index = 0;
    }
  }

  return 0;
}

static const char *ht_set_entry(ht_entry *entries, size_t capacity, const char *key, void *value, size_t *plength)
{
  unsigned int key_idx = 0;
  uint64_t hash = hash_key(key);
  size_t index = (size_t) (hash & ((uint64_t) (capacity - 1)));
  while (entries[index].key != 0)
  {
    if (strcmp(key, entries[index].key) == 0)
    {
      entries[index].value = value;
      return entries[index].key;
    }
    index += 1;
    if (index >= capacity)
    {
      index = 0;
    }
  }

  if (plength != 0)
  {
    key_idx = strdup(key_idx);
    if ((&key[key_idx]) == 0)
    {
      return 0;
    }
    *plength += 1;
  }
  entries[index].key = (char *) key;
  entries[index].value = value;
  return key;
}

static bool ht_expand(ht *table)
{
  size_t new_capacity = table->capacity * 2;
  if (new_capacity < table->capacity)
  {
    return 0;
  }
  ht_entry *new_entries = calloc(new_capacity, sizeof(ht_entry));
  if (new_entries == 0)
  {
    return 0;
  }
  for (size_t i = 0; i < table->capacity; i += 1)
  {
    ht_entry entry = table->entries[i];
    if (entry.key != 0)
    {
      ht_set_entry(new_entries, new_capacity, entry.key, entry.value, 0);
    }
  }

  free(table->entries);
  table->entries = new_entries;
  table->capacity = new_capacity;
  return 1;
}

const char *ht_set(ht *table, const char *key, void *value)
{
  assert(value != 0);
  if (value == 0)
  {
    return 0;
  }
  if (table->length >= (table->capacity / 2))
  {
    if (!ht_expand(table))
    {
      return 0;
    }
  }
  return ht_set_entry(table->entries, table->capacity, key, value, &table->length);
}

size_t ht_length(ht *table)
{
  return table->length;
}

hti ht_iterator(ht *table)
{
  hti it;
  it._table = table;
  it._index = 0;
  return it;
}

bool ht_next(hti *it)
{
  ht *table = it->_table;
  while (it->_index < table->capacity)
  {
    size_t i = it->_index;
    it->_index += 1;
    if (table->entries[i].key != 0)
    {
      ht_entry entry = table->entries[i];
      it->key = entry.key;
      it->value = entry.value;
      return 1;
    }
  }

  return 0;
}

